compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/optimizer/BIROptimizer.java

```diff
@@ -312,86 +316,70 @@ private void reuseTempVariables(BIRFunction birFunction, OptimizerEnv funcOpEnv)
+                            k -> new LinkedHashSet<>());
+                    if (reusableList.isEmpty()) {
+                        // No variables to reuse hence continue.
+                        continue;
                     }
+                    // We always get oldest reusable var.
+                    BIRVariableDcl reusableVar = reusableList.iterator().next();
+                    removableVarVsReplacementVarMap.put(variableDcl, reusableVar);
+                    removableOpsMap.put(birOperand, reusableVar);
+                    reusableList.remove(reusableVar);
+                    reusableList.remove(variableDcl);
+                    replaceableVarSet.add(reusableVar);
                     continue;
                 }
-                secondLoadedSet.add(variableDcl);
                 BIRVariableDcl reusableVar = removableVarVsReplacementVarMap.get(variableDcl);
-                // Same var can be used by multiple operands. we need to collect all and then replace the it
+                Set<BIRVariableDcl> reusableList = typeVsReusableVarsMap.get(variableDcl.type);
+                // Same var can be used by multiple operands. We need to collect all and then replace the it
                 // variable declarations.
                 if (reusableVar != null) {
                     removableOpsMap.put(birOperand, reusableVar);
+                    // Now previous reusable var can be reused.
+                    reusableList.add(reusableVar);
+                    continue;
                 }
+                // variable can be reused since it is used last time.
+                reusableList.add(variableDcl);
             }
             // Replace vars with removable vars.
             removableOpsMap.forEach((birOperand, birVariableDcl) -> birOperand.variableDcl = birVariableDcl);
-            birFunction.localVars.removeAll(removableVarVsReplacementVarMap.keySet());
+            localVars.removeAll(removableVarVsReplacementVarMap.keySet());
+        }
+
+        private void optimizeFrameVars(List<BIRVariableDcl> localVars,
+                                       Map<BIRVariableDcl, BIRBasicBlock> basicBlockTempVars,
+                                       Set<BIRVariableDcl> replaceableVarSet) {
+            basicBlockTempVars.keySet().removeAll(replaceableVarSet);
+            for (BIRVariableDcl birVariableDcl : localVars) {
+                if (basicBlockTempVars.containsKey(birVariableDcl)) {
+                    birVariableDcl.onlyUsedInSingleBB = true;
+                }
+            }
         }
 
         @Override
         public void visit(BIRNode.BIRBasicBlock birBasicBlock) {
+            this.env.currentBB = birBasicBlock;
             this.env.newInstructions = new ArrayList<>();
             birBasicBlock.instructions.forEach(i -> this.optimizeNonTerm(i, this.env));
             birBasicBlock.instructions = this.env.newInstructions;
+            this.env.isTerminator = true;
```

Comment:
Shall we move this assignment inside the `optimizeTerm` method? Changing it before & after the `optimizeTerm()` call seems a bit unclear.